Refactoring Handlers and Business Logic, Middleware into Service Architecture

pwdel opened the pull request:
## BACKGROUND

* Refactoring `handlers → domain service → repository` is will allow us to lift `markets/` into its own microservice with minimal churn.
* To split it out, after we build the domain+repo for all needed parts, we just replace one adapter in the monolith with a thin HTTP client and move the current domain+repo into a new service process.

### HIGH LEVEL OVERVIEW

#### Markets Flow (new layered architecture)

```mermaid
flowchart TB
    client["Client"]
    handlers["handlers/markets/*.go<br/>• JSON ⇄ DTO mapping<br/>• call service interface"]
    dto["handlers/markets/dto<br/>• DTO definitions"]
    container["internal/app/container<br/>• wires handler → service → repository"]
    domain["internal/domain/markets<br/>• business logic<br/>• validation & orchestration"]
    repo["internal/repository/markets<br/>• persistence (GORM)"]
    db["Database"]

    client --> handlers
    handlers --> dto
    dto --> container
    container --> domain
    domain --> repo
    repo --> db
```

#### Markets Flow (Legacy Architecture)

```
Handlers: HTTP-only glue (JSON in/out).
Domain internal/domain/markets: business API ⇒ the “service contract”.
Repository internal/repository/markets: persistence adapter (GORM).
Container: composes repo → service → handler.
```

```mermaid
flowchart TB
    client["Client"]
    handlers["handlers/bets/*.go<br/>• request parsing<br/>• business logic<br/>• direct util.GetDB access"]
    db["Database"]

    client --> handlers --> db
```

#### Execution Plan

* Where does this get us in terms of migrating to Microservices?

[Execution Plan for Migration to Microservices](https://github.com/openpredictionmarkets/socialpredict/pull/581#issuecomment-3428961967)

## 💡 __ARCHITECTURAL STRUCTURE__ - PER DIRECTORY, PART OF THE CODE

### __OVERVIEW:__

- ✅ __Complete Domain Layer__ - All business logic patterns established
- ✅ __Repository Pattern__ - Data access cleanly separated
- ✅ __DTO Pattern__ - HTTP contracts separate from domain models
- ✅ __Service Interfaces__ - Clean dependency injection ready
- ✅ __Error Mapping__ - Domain errors → HTTP status codes
- ✅ __Build Integrity__ - System compiles and functions

### __Microservices Readiness:__

The architecture now supports __easy service extraction__:

```go
// Current: Direct repository
svc := domain.NewService(gormRepo, userSvc, clock, config)

// Future: Remote service client  
svc := domain.NewService(httpClientRepo, userSvc, clock, config)
```

---

## 📋 __COMPLETION GUIDE__

For each remaining handler file, follow this __proven pattern__:

1. __Import Changes:__

   ```go
   // Remove:
   import "socialpredict/models"
   import "gorm.io/gorm"

   // Add:
   import "socialpredict/handlers/markets/dto"
   import dmarkets "socialpredict/internal/domain/markets"
   ```

2. __Handler Signature:__

   ```go
   // Old: Direct dependencies
   func Handler(w http.ResponseWriter, r *http.Request)

   // New: Service injection  
   func NewHandler(svc dmarkets.Service) *Handler
   ```

3. __Logic Movement:__

   ```go
   // Move validation → domain.Service methods
   // Move DB calls → repository.Repository methods  
   // Keep HTTP parsing/response → handler
   ```

---

## EXAMPLE `createmarket.go` PATTERN

### __✅ FULLY COMPLETED HANDLER__

__`createmarket.go` - COMPLETE REFACTOR ✅__

- ❌ __Before__: Mixed HTTP + business logic + GORM + models
- ✅ __After__: HTTP-only handler using domain service + DTOs
- ✅ __GORM import__: Removed completely
- ✅ __Models import__: Removed completely
- ✅ __DTOs__: Uses `dto.CreateMarketRequest` → `dto.CreateMarketResponse`
- ✅ __Domain service__: Calls `dmarkets.Service.CreateMarket()`
- ✅ __Error mapping__: Domain errors → HTTP status codes

### __GUARD CHECK PROGRESS__

```bash
# Before refactoring:
GORM imports remaining: 3
Models imports remaining: 7

# After refactoring:
GORM imports remaining: 3 (unchanged - focused on one file)
Models imports remaining: 6 (down by 1 - createmarket.go ✅)
```

---

## 🏗️ __ARCHITECTURAL PATTERN PROVEN__

### __Complete Clean Handler Pattern (createmarket.go):__

__1. HTTP-Only Handler:__

```go
func (h *CreateMarketHandler) Handle(w http.ResponseWriter, r *http.Request) {
    // 1. Parse request DTO
    var req dto.CreateMarketRequest
    json.NewDecoder(r.Body).Decode(&req)
    
    // 2. Convert to domain request
    domainReq := dmarkets.MarketCreateRequest{...}
    
    // 3. Call domain service
    market, err := h.svc.CreateMarket(ctx, domainReq, username)
    
    // 4. Map errors to HTTP status
    switch err { case dmarkets.ErrUserNotFound: ... }
    
    // 5. Convert to response DTO
    response := dto.CreateMarketResponse{...}
    json.NewEncoder(w).Encode(response)
}
```

__2. Domain Service (business logic):__

```go
func (s *Service) CreateMarket(ctx, req, username) (*Market, error) {
    // All validation, orchestration, business rules
    // No HTTP knowledge, no GORM imports
}
```

__3. Repository (data access):__

```go
func (r *Repository) Create(ctx context.Context, m *Market) error {
    // GORM operations, database access
    // No business logic, no HTTP
}
```

---

##

### __FILES STATUS__

#### MARKETS

| File / Area                                   | Current State                                                                      | Status   | Notes & Next Action |
|-----------------------------------------------|--------------------------------------------------------------------------------------|----------|---------------------|
| handlers/markets/handler.go                   | Uses injected domain + auth services; no direct DB access                           | ✅ Done  | Legacy struct is clean; can eventually retire once standalone handlers cover every route. |
| handlers/markets/createmarket.go              | Service-backed create endpoint with AuthService authorization                       | ✅ Done  | Domain owns validation + persistence; nothing left to decommission. |
| handlers/markets/listmarkets.go               | Factory delegating to `svc.List` / `ListByStatus` with DTO mapping                  | ✅ Done  | Continue routing through this implementation; legacy wrapper can be removed when convenient. |
| handlers/markets/getmarkets.go                | Thin alias around `svc.ListMarkets` for backward compat                             | ✅ Done  | Safe to keep until routes converge. |
| handlers/markets/marketdetailshandler.go      | Service-injected details handler returning DTO                                      | ✅ Done  | Expand tests only when new response fields surface. |
| handlers/markets/marketprojectedprobability.go| Fully wired HTTP endpoint; domain `ProjectProbability` computes real projections via WPAM math | ✅ Done  | Projection math + tests live in the domain; no further handler work required. |
| handlers/markets/resolvemarket.go             | Service-injected resolve endpoint using auth façade                                 | ✅ Done  | Optional future cleanup: share auth helpers if we add more admin-only actions. |
| handlers/markets/searchmarkets.go             | Service-backed search with sanitization and DTO response                            | ✅ Done  | Legacy search code removed; nothing pending. |

#### USERS

| File / Area                                   | Current State                                                                                     | Status   | Notes & Next Action |
|-----------------------------------------------|-----------------------------------------------------------------------------------------------------|----------|---------------------|
| handlers/users/apply_transaction.go           | Legacy file removed; all balance adjustments route through `internal/domain/users.Service.ApplyTransaction` via handlers/services | ✅ Done  | Keep using the domain service so handlers never call `util.GetDB` directly. |
| handlers/users/changedescription.go           | Service-backed handler calling `svc.UpdateDescription`; auth via auth façade                      | ✅ Done  | – |
| handlers/users/changedisplayname.go           | Service-backed handler calling `svc.UpdateDisplayName`; auth via auth façade                      | ✅ Done  | – |
| handlers/users/changeemoji.go                 | Service-backed handler calling `svc.UpdateEmoji`; auth via auth façade                            | ✅ Done  | – |
| handlers/users/changepassword.go              | Service-backed handler calling `svc.ChangePassword`; hashing handled in domain                    | ✅ Done  | Domain tests cover success/error paths. |
| handlers/users/changepersonallinks.go         | Service-backed handler calling `svc.UpdatePersonalLinks`; auth via auth façade                    | ✅ Done  | – |
| handlers/users/financial.go                   | Service-backed handler returning `svc.GetUserFinancials`                                          | ✅ Done  | Covered by domain + handler tests. |
| handlers/users/listusers.go                   | Thin HTTP façade delegating to `svc.ListUserMarkets`                                              | ✅ Done  | – |
| handlers/users/privateuser/privateuser.go     | Service-backed private profile handler using auth façade                                          | ✅ Done  | Future consolidation optional if DTOs expand. |
| handlers/users/publicuser.go                  | Service-backed handler returning `svc.GetPublicUser`                                              | ✅ Done  | – |
| handlers/users/publicuser/portfolio.go        | Service-backed handler returning `svc.GetUserPortfolio`                                           | ✅ Done  | – |
| handlers/users/publicuser/publicuser.go       | Duplicate implementation removed; routes share the service-backed handler                         | ✅ Done  | – |
| handlers/users/userpositiononmarkethandler.go | Handler composes markets + users services (auth façade enforced); position math lives in domain   | ✅ Done  | – |
| handlers/users/credit/usercredit.go           | Service-backed handler returning `svc.GetUserCredit` with missing-user fallback                   | ✅ Done  | – |

#### FINANCIALS and MATH

| File / Area                                  | Current State                                                                                                    | Status   | Notes & Next Action |
|----------------------------------------------|------------------------------------------------------------------------------------------------------------------|----------|---------------------|
| financials/financialsnapshot.go              | Implemented inside `internal/domain/analytics.Service.ComputeUserFinancials`; metrics + user endpoints consume it | ✅ Done  | Ready for general use; expand load-testing only if we onboard heavier portfolios. |
| financials/systemmetrics.go                  | Served via `analytics.Service.ComputeSystemMetrics`; metrics handler injects analytics service                   | ✅ Done  | All callers switched to the service; no further action required. |
| financials/workprofits.go                    | Rebuilt behind analytics domain/service with repository + tests                                                  | ✅ Done  | Keep monitoring accuracy as new profit types are added. |
| market/dust.go                               | Pure math helper under `internal/domain/math/market`; buy/sell flows reference domain service wrappers           | ✅ Done  | No action needed; façade already shields handlers. |
| market/marketvolume.go                       | Volume calculations now invoked through the markets domain service (bets fetched via repository layer)           | ✅ Done  | Watch for performance regressions on very large bet sets. |
| outcomes/dbpm/marketshares.go                | Pure math under `internal/domain/math/outcomes/dbpm`; consumed via positions math service                        | ✅ Done  | Nothing pending. |
| payout/resolvemarketcore.go                  | Logic absorbed into `internal/domain/markets.Service.ResolveMarket`                                              | ✅ Done  | Legacy helpers removed; docs updated to match new flow. |
| positions/positionsmath.go & related modules | Wrapped by positions domain service; no direct GORM usage from handlers                                          | ✅ Done  | Further tweaks only if we add new position types. |
| positions/adjust_valuation.go                | Runs through positions service with repository dependencies injected                                             | ✅ Done  | Covered by resolve/settlement tests. |
| positions/earliest_users.go                  | Moved behind repository method with deterministic tests; exposed via positions service                           | ✅ Done  | No outstanding work. |
| probabilities/wpam/*.go                      | Pure math utilities packaged under domain math probabilities and accessed via service abstractions               | ✅ Done  | Optional future enhancement: dedicated probabilities microservice. |
| README_MATH.md                               | Updated to describe the new analytics + market service façades and math ownership                                | ✅ Done  | Keep in sync as additional math modules migrate. |

#### BETS

| File                                             | Current State                                             | Still Tied to DB/Models? | Status   | Notes / Next Action                                                                 |
|--------------------------------------------------|-----------------------------------------------------------|---------------------------|----------|---------------------------------------------------------------------------------------|
| handlers/bets/betshandler.go                     | Service-backed GET /markets/{id}/bets via markets svc     | No                        | ✅ Done  | Uses domain service output; nothing pending.                                         |
| handlers/bets/listbetshandler.go                 | Removed; behaviour now lives in markets service           | —                         | ✅ Done  | Replace references with betshandler.go wiring if any linger.                         |
| handlers/bets/buying/buypositionhandler.go       | Thin HTTP layer calling bets service                      | No                        | ✅ Done  | Domain/service tests cover validation and balance checks.                            |
| handlers/bets/selling/sellpositionhandler.go     | Thin HTTP layer calling bets service                      | No                        | ✅ Done  | Service exposes dust-cap errors; handler only maps responses.                        |
| handlers/bets/selling/sellpositioncore.go        | Deleted during sell-flow migration                        | —                         | ✅ Done  | Dust/share math now in bets.Service.calculateSale.                                   |
| handlers/bets/betutils/*.go                      | Deleted legacy DB helpers                                 | —                         | ✅ Done  | Validation handled by bets domain service/repo.                                      |
| handlers/bets/selling/sellingdust.go             | Deleted stub                                              | —                         | ✅ Done  | Dust reporting handled by bets service & DTOs.                                       |
| handlers/bets/errors.go                          | Deleted; ErrDustCapExceeded lives in bets domain          | —                         | ✅ Done  | Handler asserts against domain error type.                                           |
| handlers/bets/market_status_validation_test.go   | Deleted with legacy helpers                               | —                         | ✅ Done  | Covered by bets service tests (e.g., market closed).                                 |
| handlers/bets/listbetshandler.go (tests)         | Deleted with handler removal                              | —                         | ✅ Done  | Market bet history now tested in service_marketbets_test.go.                         |

#### MIDDLEWARE

| File / Area                                   | Current State                                                                 | Status   | Notes & Next Action                                                                                     |
|-----------------------------------------------|-------------------------------------------------------------------------------|----------|----------------------------------------------------------------------------------------------------------|
| middleware/auth_legacy.go                     | File removed; DB-backed helper eliminated                                    | ✅ Done  | Nothing left to migrate                                                                                 |
| internal/service/auth (formerly middleware)   | Auth façade with `Authenticator` interface and `AuthService` in place        | ✅ Done  | Continue using as canonical API for authentication/authorization                                        |
| handlers/markets/handler.go                   | Uses injected `AuthService` facade                                           | ✅ Done  | No further work                                                                                         |
| handlers/markets/createmarket.go              | Auth now routed through `AuthService`                                       | ✅ Done  | —                                                                                                       |
| Other handlers (admin, etc.)                  | All migrated to service-based auth helpers                                   | ✅ Done  | Continue to inject `AuthService` for any new handlers                                                    |
| Container wiring (`internal/app`)             | Builds and exposes `AuthService`; handlers receive it via DI                 | ✅ Done  | —                                                                                                       |
| Middleware packaging (legacy)                 | Migrated under `internal/service/auth`; mixed responsibilities resolved      | ✅ Done  | Document new location/usage if needed                                                                   |


#### OpenAPI Update

* What was previously merely a server based system is now a contract-based system using OpenAPI.
* Documentation can be found in openapi.yaml in backend/
* Below is an accounting of all routes in the server and openapi.yaml as well as anything deprecated.

| Route | In `server.go` | In `docs/openapi.yaml` | Notes |
| --- | --- | --- | --- |
| GET /v0/home | ✅ | ✅ | – |
| POST /v0/login | ✅ | ✅ | – |
| GET /v0/setup | ✅ | ✅ | – |
| GET /v0/stats | ✅ | ✅ | – |
| GET /v0/system/metrics | ✅ | ✅ | – |
| GET /v0/global/leaderboard | ✅ | ✅ | – |
| GET /v0/markets | ✅ | ✅ | – |
| GET /v0/markets/search | ✅ | ✅ | – |
| GET /v0/markets/active | ✅ | ✅ | – |
| GET /v0/markets/closed | ✅ | ✅ | – |
| GET /v0/markets/resolved | ✅ | ✅ | – |
| GET /v0/markets/{id} | ✅ | ✅ | – |
| GET /v0/marketprojection/{marketId}/{amount}/{outcome} | ✅ | ✅ | – |
| GET /v0/markets/bets/{marketId} | ✅ | ✅ | – |
| GET /v0/markets/positions/{marketId} | ✅ | ✅ | – |
| GET /v0/markets/positions/{marketId}/{username} | ✅ | ✅ | – |
| GET /v0/markets/{id}/leaderboard | ✅ | ✅ | – |
| GET /v0/userinfo/{username} | ✅ | ✅ | – |
| GET /v0/usercredit/{username} | ✅ | ✅ | – |
| GET /v0/portfolio/{username} | ✅ | ✅ | – |
| GET /v0/users/{username}/financial | ✅ | ✅ | – |
| GET /v0/privateprofile | ✅ | ✅ | – |
| POST /v0/changepassword | ✅ | ✅ | – |
| POST /v0/profilechange/displayname | ✅ | ✅ | – |
| POST /v0/profilechange/emoji | ✅ | ✅ | – |
| POST /v0/profilechange/description | ✅ | ✅ | – |
| POST /v0/profilechange/links | ✅ | ✅ | – |
| POST /v0/markets/{id}/resolve | ✅ | ✅ | Canonical route; frontend now uses this path. |
| POST /v0/bet | ✅ | ✅ | – |
| GET /v0/userposition/{marketId} | ✅ | ✅ | – |
| POST /v0/sell | ✅ | ✅ | – |
| POST /v0/create | ✅ | ✅ | Legacy alias to POST /v0/markets. |
| POST /v0/admin/createuser | ✅ | ✅ | – |
| GET /v0/content/home | ✅ | ✅ | – |
| PUT /v0/admin/content/home | ✅ | ✅ | – |

pwdel committed:
Updating refactor of markets handler.

pwdel committed:
Adding working markets.

pwdel committed:
Updating internal.

pwdel committed:
Updating service

pwdel committed:
Updating and moving createmarket and responses.

pwdel committed:
Fixed tests.

pwdel committed:
Refactoring listmarkets.go

pwdel committed:
Updating listmarketsbystatus.go

pwdel committed:
All tests passing.

pwdel committed:
Updating, refactoring all market operations as service with service injection in the server.

pwdel committed:
Merge branch 'main' into fix/checkpoint20251020-80

pwdel committed:
Refactoring further.

pwdel committed:
Update, attempting further fixes.

pwdel committed:
Merge branch 'main' into fix/checkpoint20251020-80

pwdel committed:
Update tests passing.

pwdel committed:
Adding test coverage for user functions.

pwdel committed:
Adding container test.

pwdel committed:
Updating further migrations.

pwdel committed:
Updating.

pwdel committed:
Merge branch 'main' into fix/checkpoint20251020-80

pwdel committed:
Migrating users to service domain

pwdel committed:
Updating middleware, handlers, math directory and auth directory.

pwdel committed:
Update.

pwdel committed:
Migrating bets

pwdel committed:
Updating math domain positions, market models, container, position test.

pwdel committed:
Update.

pwdel committed:
Refactoring position math so that it no longer touches GORM, using input type onf market MarketSnapshot to take in bet results.

pwdel committed:
Updating

pwdel committed:
Updating per codeql recommendations.

pwdel committed:
Adding OpenAPI spec and starting to test payloads

pwdel committed:
Adjusting openapi to describe payload for /v0/markets/{id} more correctly.

pwdel committed:
Fixing market display page.

pwdel committed:
Fixing initial user balance.

pwdel committed:
Updaitng buyshares to fit with openAPI layout.

pwdel committed:
Taking non-needed logging out of the frontend.

pwdel committed:
Adding positions back in after reformatting.

pwdel committed:
Updating to fix leaderboards.

pwdel committed:
Updating, fixing selling shares.

pwdel committed:
Update API camelCase

pwdel committed:
Updating, attempting to fix search

pwdel committed:
Bringing openapi up to date with what has been updated.

pwdel committed:
Updating.

pwdel committed:
Updating and fixing search.

pwdel committed:
Updating, attempting to resolve.

pwdel committed:
Adjusting status sigfigs to 2

astrosnat commented:
testing on localhost...
<img width="1300" height="736" alt="image" src="https://github.com/user-attachments/assets/85151029-e7d9-49d2-ade4-3558fffdf723" />
3 significant figures, not 2, but the important part is this works

<img width="1300" height="736" alt="image" src="https://github.com/user-attachments/assets/e7ccf171-ba88-4e6d-a064-f75af26260a3" />
this is an unclear error message when trying to bet on a resolved market, but this is a nitpick

overall, looks good to me, happy to approve if you have tested this on kconfs @pwdel

pwdel committed:
Update with GIF

pwdel committed:
Reverting back.

pwdel committed:
Merge branch 'main' into fix/checkpoint20251020-80

raisch commented:
There so much here, rather than producing a full review, I'll make some general comments here.

When I review another engineer's work, I try to look at everything as holistically as I can, using standard development Best Practices as my polestar. 

I look specifically for things like 
- How complex the code is ("cyclomatic complexity") which makes the code difficult to reason about and maintain.
- Violations of the Open/Closed Principle that adversely affect maintainability but also tend to make code very brittle, such that a small change in one place has unattended affect in other areas of the codebase. (This is also known as "coupling", as in how we strive for "low coupling/high cohesion".)
- Violations of the Single Responsibility Principle which cause all of the above issues, but I think most importantly, makes it very hard to reason about program flow. (I have a whole lecture I give about how our minds can only deal with 7 - 9 things at any one time due to how we process short-term vs. long-term memory.)

The goal of this evaluation is to help make the codebase:
- Reliable
- Predictable
- Easy to reason about
- Easier to modify without unexpected consequences
- Easier to maintain by a team.

Note that of the worst offenders listed below, most of the files/functions appear in multiple categories.

## Cyclomatic Complexity

An old but I feel still pertinent measure of how many paths there are out of a function/method/etc. I view anything above 8 to be unnecessarily complex. (In my own work, I strive for 4 or less.)

Worst offenders:

- internal/domain/analytics/service.go (line 241) ComputeGlobalLeaderboard has cyclomatic **20**: pulls markets/bets, calculates user aggregates, determines earliest bets, and sorts/ranks in one method. Splitting data access, aggregation, and ranking into helpers would satisfy SRP and bring complexity under the ≤8 target.
- internal/domain/analytics/service.go (line 89) ComputeSystemMetrics (cyclomatic **14**) performs debt arithmetic, market fee math, participation fee counting, and response construction in a single block while issuing N+1 market bet queries. Extract per-concern calculators (user debt, active volume, fee tallies) and encapsulate repository calls to reduce complexity and improve testability.
- handlers/markets/searchmarkets.go (line 15) SearchMarketsHandler (cyclomatic **19**) mixes HTTP method checks, query parsing, validation/sanitization, pagination defaults, domain calls, DTO mapping, and response encoding. A request parser + responder layer would lower complexity and align with Single Responsibility.
- handlers/markets/listmarkets.go (line 14) ListMarketsHandlerFactory (cyclomatic **15**) similarly combines parameter parsing, branching between services, error mapping, DTO building, and encoding. Extracting pagination/status parsing and response writing into dedicated helpers/middleware would shrink the handler’s complexity.
- internal/domain/markets/service.go (line 370) SearchMarkets and  (line 444) ResolveMarket (cyclomatic **15**/**16**) both bundle validation, fallback search logic, and payout/refund workflows. Consider moving payout transaction loops to a transaction-oriented collaborator and isolating fallback-search assembly to dedicated functions to hit the complexity ≤8 goal and keep domain operations single-purpose.
- util/postgres.go (line 15) InitDB (cyclomatic **10**) owns env resolution, DSN construction, connection open, and global mutation, and aborts the process on failure. Returning an error and encapsulating config parsing separately would drop complexity and better respect SOLID (no hidden global state).

## Single Responsibility Principle

"A function/module/etc. should only have one reason to change."

The example I use to describe this is a function that retrieves data from some source, transforms it into printable form, and sends it to a printer.

In this function, there are many "reasons to change": if
- the function needs to pull its data from somewhere else
- the transformation (formatting) needs to be updated
- the result needs to be sent to the file system, external url, etc. rather than the printer

Three reasons to change, in one function. (This is related to Open/Closed below.)

There are lots of violations, but the worst offenders are:

- internal/domain/analytics/service.go (line 241) ComputeGlobalLeaderboard — data fetching, per-market aggregation, per-user aggregation, earliest-bet detection, sorting, and ranking all in one routine.
- internal/domain/analytics/service.go (line 89) ComputeSystemMetrics — mixes config access, user debt math, market fee/volume tallying, participation fee detection, and response assembly plus multiple repo queries.
- handlers/markets/searchmarkets.go (line 15) SearchMarketsHandler — HTTP method gate, query parsing, validation/sanitization, paging defaults, domain calls, DTO mapping, and JSON encoding in one function.
- handlers/markets/listmarkets.go (line 14) ListMarketsHandlerFactory — parameter parsing, branching service calls, error mapping, DTO building, and response writing combined.
- internal/domain/markets/service.go (line 444) ResolveMarket — input validation, auth checks, repo updates, and separate refund/payout transaction loops bundled together.

## Open/Closed Principle

"A function/module/etc. should be open to extension but closed to modification."

Any time you need to go back into the code to fix a bug, without strict adherence to this principle, you **will** cause more bugs to pop up based on your modifications to existing code. (This is another instance of "coupling".)

The best way to ameliorate this problem is to "lock" existing work and use extension to apply modifications.

The worst offenders:

- internal/domain/analytics/service.go (line 89) ComputeSystemMetrics — tightly bundles debt, fee, and volume calculations with data access; any new fee type or adjustment requires modifying this method rather than extending via strategy/helpers.
- internal/domain/analytics/service.go (line 241) ComputeGlobalLeaderboard — aggregation, earliest-bet detection, and ranking are hard-coded; adding a new ranking dimension or tie-breaker forces editing the core function instead of composing strategies.
- handlers/markets/searchmarkets.go (line 15) SearchMarketsHandler — request parsing, validation/sanitization, domain call, and response formatting are intertwined; adding new filters or response fields means touching this handler rather than plugging in middleware or mappers.
- handlers/markets/listmarkets.go (line 14) ListMarketsHandlerFactory — same pattern: branching logic and DTO assembly live in one place, so new list variants/status handling require edits instead of extension.
- internal/domain/markets/service.go (line 370) SearchMarkets — fallback search logic and filtering rules are embedded; changing thresholds, statuses, or fallback strategy demands direct modification.
- internal/domain/markets/service.go (line 444) ResolveMarket — payout/refund pathways are inlined; adding new resolutions or payout rules requires changing this method rather than supplying pluggable settlement strategies.

## Other Violations of SOLID Principles

- util/postgres.go (line 15) InitDB — violates SRP and Dependency Inversion: hides configuration parsing, connection building, and global state mutation; callers can’t inject DB, forcing hard coupling and fatal exit paths.
- internal/service/auth/loggin.go (line 28) LoginHandler — high coupling and DIP/LSP issues: directly depends on global DB via util.GetDB and constructs SecurityService inline; no interface boundaries make swapping storage/auth strategies or testing difficult.
- internal/domain/analytics/service.go (repo field is a concrete interface but still) — hidden N+1 queries in ComputeSystemMetrics/ComputeGlobalLeaderboard break Interface Segregation/Dependency Inversion because Repository forces both ordered and per-market bet calls; no separate query abstractions for bulk operations.
- handlers/markets/searchmarkets.go and handlers/markets/listmarkets.go — handlers depend on concrete svc behaviors and perform validation/sanitization/mapping themselves, mixing concerns and resisting Open/Closed; lack of small interfaces for parsing/response writing impedes substitution/testing.
- internal/domain/markets/service.go (line 444) ResolveMarket — violates SRP/OCP: service controls auth, settlement state, repo updates, and user transactions directly; no strategy or collaborator to extend payout/refund logic without modifying the method.
- security/sanitizer.go (line 292) isValidEmoji — LSP/ISP risk: “sanitizer” both validates and parses; callers can’t substitute a different emoji/allow-list without reworking the sanitizer, and the function accepts any ASCII, weakening contract clarity.

## Comments on Coupling/Cohesion

Coupling is how strongly one module depends on another’s details. High coupling means changes in one piece force changes elsewhere (shared state, concrete types, hidden assumptions); low coupling keeps modules talking through stable abstractions so they can evolve independently.

Cohesion is how tightly the responsibilities within a module relate to a single purpose. High cohesion means the code in a unit works toward one focused task; low cohesion means a grab bag of unrelated duties living together, making the unit harder to understand, test, and change.

Good engineers alway strive for Low Coupling and High Cohesion.

pwdel committed:
Attempting reduction in cyclomatic complexity.

astrosnat committed:
Merge branch 'main' into fix/checkpoint20251020-80

pwdel committed:
Single responsibility improvement of calculate metrics.

pwdel committed:
Single responsibility fix.

raisch reviewed (commented):
`docs/openapi.yaml` is missing five routes: 

- `GET /health`
- `GET /v0/setup/frontend`
- `GET /v0/global/leaderboard`
- `GET /v0/markets/status/{status}` - see note below.
- `GET /v0/markets/{id}/projection`

And still documents four legacy shortcuts: 

- `GET /v0/markets/active`
- `GET /v0/markets/closed`
- `GET /v0/markets/resolved`
- `GET /v0/marketprojection/{marketId}/{amount}/{outcome}`
  - the code now uses `GET /v0/markets/{id}/projection` instead
 
Note: 

Re: `GET /v0/markets/status/{status}` - accepts `status` values of "active", "closed", "resolved", "all" but the underlying code accepts "" as well. If this is intended, you should also have a `GET /v0/markets/status` route as well and drop the "all" status.

raisch committed:
added /v0/market/status as a shim over /v0/market/status/{status}, setting status = "all"

raisch committed:
updated per code

raisch committed:
added /v0/openapi.yaml to serve spec

raisch committed:
embedded swagger in backend; updated docs; still more to do there tho.

raisch committed:
removed comments

raisch committed:
fix 500 resp when market resolution time is too short.

raisch committed:
added analysis of API, updated openapi.yaml

raisch committed:
updated

pwdel committed:
Attempted refactor system metrics towards OCP

pwdel committed:
Updating per no inverted dependencies principle

pwdel committed:
Reducing cyclomatic complexity within security sanitizer

pwdel committed:
Updating, targeting lower OCP on SearchMarketsHandler.

pwdel committed:
Updating handler, service for cyclomatic complexity reduction.

pwdel committed:
Adding gocache dir to gitignore to prevent super long git operations.

pwdel committed:
Updating for cyclomatic complexity.

pwdel committed:
Lowering all non-test functions to cyclomatic complexity less than 8

pwdel committed:
Reducing cyclomatic complexity.

pwdel commented:
Reduced cyclomatic complexity on all non-test functions to less than 9.

pwdel committed:
Refactored bets/service.go to make each function a single concern, simplified place/sell into focused helpers.

pwdel committed:
Adding focused unit tests to bets service domain.

astrosnat committed:
initial redesign draft for /users/service.go (#603)

* Add checkpoint for git commit timing experiment

* Adding SRP explainer checkpoint.

This brief description of the Single Responsibility Principle (SRP) is intended for "vibe coding" tools like ChatGPT Codex to read through and use as a basis for evaluating and generating code.

* Codex evaluation checkpoint

Codex evaluation of SRP adherence in CMS service.go

* Evaluate SRP adherence

Evaluating SRP adherence of `/internal/domain/users/service.go` using Codex

* initial redesign for /users/service.go?

pwdel commented:
OK, starting project management for remaining tasks:

### SRP
| Area | File | Assignee | Status |
| --- | --- | --- | --- |
| analytics | backend/internal/domain/analytics/financialsnapshot_test.go | Patrick | [ ] |
| analytics | backend/internal/domain/analytics/leaderboard_test.go | Patrick | [ ] |
| analytics | backend/internal/domain/analytics/models.go | Patrick | [ ] |
| analytics | backend/internal/domain/analytics/repository.go | Patrick | [ ] |
| analytics | backend/internal/domain/analytics/service.go | Patrick | [ ] |
| analytics | backend/internal/domain/analytics/systemmetrics_integration_test.go | Patrick | [ ] |
| analytics | backend/internal/domain/analytics/systemmetrics_test.go | Patrick | [ ] |
| bets | backend/internal/domain/bets/errors.go | Patrick | [ ] |
| bets | backend/internal/domain/bets/models.go | Patrick | [ ] |
| bets | backend/internal/domain/bets/service_helpers_test.go | Patrick | [ ] |
| bets | backend/internal/domain/bets/service_test.go | Patrick | [ ] |
| bets | backend/internal/domain/bets/service.go | Patrick | [ ] |
| markets | backend/internal/domain/markets/errors.go | Patrick | [ ] |
| markets | backend/internal/domain/markets/models.go | Patrick | [ ] |
| markets | backend/internal/domain/markets/service_details_test.go | Patrick | [ ] |
| markets | backend/internal/domain/markets/service_listbystatus_test.go | Patrick | [ ] |
| markets | backend/internal/domain/markets/service_marketbets_test.go | Patrick | [ ] |
| markets | backend/internal/domain/markets/service_probability_test.go | Patrick | [ ] |
| markets | backend/internal/domain/markets/service_resolve_test.go | Patrick | [ ] |
| markets | backend/internal/domain/markets/service_search_test.go | Patrick | [ ] |
| markets | backend/internal/domain/markets/service.go | Patrick | [ ] |
| math | backend/internal/domain/math/market/dust_test.go | Patrick | [ ] |
| math | backend/internal/domain/math/market/dust.go | Patrick | [ ] |
| math | backend/internal/domain/math/market/marketvolume_test.go | Patrick | [ ] |
| math | backend/internal/domain/math/market/marketvolume.go | Patrick | [ ] |
| math | backend/internal/domain/math/outcomes/dbpm/marketshares_test.go | Patrick | [ ] |
| math | backend/internal/domain/math/outcomes/dbpm/marketshares.go | Patrick | [ ] |
| math | backend/internal/domain/math/positions/adjust_valuation_test.go | Patrick | [ ] |
| math | backend/internal/domain/math/positions/adjust_valuation.go | Patrick | [ ] |
| math | backend/internal/domain/math/positions/positionsmath_test.go | Patrick | [ ] |
| math | backend/internal/domain/math/positions/positionsmath.go | Patrick | [ ] |
| math | backend/internal/domain/math/positions/profitability_test.go | Patrick | [ ] |
| math | backend/internal/domain/math/positions/profitability.go | Patrick | [ ] |
| math | backend/internal/domain/math/positions/valuation_test.go | Patrick | [ ] |
| math | backend/internal/domain/math/positions/valuation.go | Patrick | [ ] |
| math | backend/internal/domain/math/probabilities/wpam/wpam_current_test.go | Patrick | [ ] |
| math | backend/internal/domain/math/probabilities/wpam/wpam_current.go | Patrick | [ ] |
| math | backend/internal/domain/math/probabilities/wpam/wpam_marketprobabilities_test.go | Patrick | [ ] |
| math | backend/internal/domain/math/probabilities/wpam/wpam_marketprobabilities.go | Patrick | [ ] |
| users | backend/internal/domain/users/errors.go | Osnat | [ ] |
| users | backend/internal/domain/users/models.go | Osnat | [ ] |
| users | backend/internal/domain/users/service_profile_test.go | Osnat | [ ] |
| users | backend/internal/domain/users/service_transactions_test.go | Osnat | [ ] |
| users | backend/internal/domain/users/service.go | Osnat | [ ] |
| users | backend/internal/domain/users/transactions.go | Osnat | [ ] |

pwdel commented:
### ISP
| Area | File | Assignee | Status |
| --- | --- | --- | --- |
| analytics | backend/internal/domain/analytics/financialsnapshot_test.go | Patrick | [ ] |
| analytics | backend/internal/domain/analytics/leaderboard_test.go | Patrick | [ ] |
| analytics | backend/internal/domain/analytics/models.go | Patrick | [ ] |
| analytics | backend/internal/domain/analytics/repository.go | Patrick | [ ] |
| analytics | backend/internal/domain/analytics/service.go | Patrick | [ ] |
| analytics | backend/internal/domain/analytics/systemmetrics_integration_test.go | Patrick | [ ] |
| analytics | backend/internal/domain/analytics/systemmetrics_test.go | Patrick | [ ] |
| bets | backend/internal/domain/bets/errors.go | Patrick | [ ] |
| bets | backend/internal/domain/bets/models.go | Patrick | [ ] |
| bets | backend/internal/domain/bets/service_helpers_test.go | Patrick | [ ] |
| bets | backend/internal/domain/bets/service_test.go | Patrick | [ ] |
| bets | backend/internal/domain/bets/service.go | Patrick | [ ] |
| markets | backend/internal/domain/markets/errors.go | Patrick | [ ] |
| markets | backend/internal/domain/markets/models.go | Patrick | [ ] |
| markets | backend/internal/domain/markets/service_details_test.go | Patrick | [ ] |
| markets | backend/internal/domain/markets/service_listbystatus_test.go | Patrick | [ ] |
| markets | backend/internal/domain/markets/service_marketbets_test.go | Patrick | [ ] |
| markets | backend/internal/domain/markets/service_probability_test.go | Patrick | [ ] |
| markets | backend/internal/domain/markets/service_resolve_test.go | Patrick | [ ] |
| markets | backend/internal/domain/markets/service_search_test.go | Patrick | [ ] |
| markets | backend/internal/domain/markets/service.go | Patrick | [ ] |
| math | backend/internal/domain/math/market/dust_test.go | Patrick | [ ] |
| math | backend/internal/domain/math/market/dust.go | Patrick | [ ] |
| math | backend/internal/domain/math/market/marketvolume_test.go | Patrick | [ ] |
| math | backend/internal/domain/math/market/marketvolume.go | Patrick | [ ] |
| math | backend/internal/domain/math/outcomes/dbpm/marketshares_test.go | Patrick | [ ] |
| math | backend/internal/domain/math/outcomes/dbpm/marketshares.go | Patrick | [ ] |
| math | backend/internal/domain/math/positions/adjust_valuation_test.go | Patrick | [ ] |
| math | backend/internal/domain/math/positions/adjust_valuation.go | Patrick | [ ] |
| math | backend/internal/domain/math/positions/positionsmath_test.go | Patrick | [ ] |
| math | backend/internal/domain/math/positions/positionsmath.go | Patrick | [ ] |
| math | backend/internal/domain/math/positions/profitability_test.go | Patrick | [ ] |
| math | backend/internal/domain/math/positions/profitability.go | Patrick | [ ] |
| math | backend/internal/domain/math/positions/valuation_test.go | Patrick | [ ] |
| math | backend/internal/domain/math/positions/valuation.go | Patrick | [ ] |
| math | backend/internal/domain/math/probabilities/wpam/wpam_current_test.go | Patrick | [ ] |
| math | backend/internal/domain/math/probabilities/wpam/wpam_current.go | Patrick | [ ] |
| math | backend/internal/domain/math/probabilities/wpam/wpam_marketprobabilities_test.go | Patrick | [ ] |
| math | backend/internal/domain/math/probabilities/wpam/wpam_marketprobabilities.go | Patrick | [ ] |
| users | backend/internal/domain/users/errors.go | Osnat | [ ] |
| users | backend/internal/domain/users/models.go | Osnat | [ ] |
| users | backend/internal/domain/users/service_profile_test.go | Osnat | [ ] |
| users | backend/internal/domain/users/service_transactions_test.go | Osnat | [ ] |
| users | backend/internal/domain/users/service.go | Osnat | [ ] |
| users | backend/internal/domain/users/transactions.go | Osnat | [ ] |

pwdel committed:
SRP for bets, markets and math. (#608)

* SRP for bets, markets and math.

* Updating with passed tests.

* Updating SRP for financials, leaderboard, analytics

* Simplifying positionsmath, valuation, wpam

* SRP on selldusst calculator.

---------

Co-authored-by: Patrick W. Delaney <patrick@Mac.hsd1.tx.comcast.net>

pwdel committed:
Updating Open Closed Principle for Various (#610)

* Updating OCP for bets

* Refactoring markets for OCP

* Expanded dust.go to allow custom sell dust policies, include ProbabilityProvider, PayoutModel, PositionCalculator to encapsulate strategies, existing APIs now delegate through calculator, default WPAM-DBPM implementations wired for new interfaces.

* Added MarketPositionCalculator interface with WPAM DBPM implementation, wired strategy injection into both analytics Service, WithPositionCalculator option, and GormRepositoroy, default calculator, nil-safety, Refactored leaderboard data loading

---------

Co-authored-by: Patrick W. Delaney <patrick@Mac.hsd1.tx.comcast.net>

pwdel committed:
Working on LSP for analytics, bets, markets, math domains (#611)

* Exported sale calculation data and switched to SaleCalculator.Calculate to return it, updating Sell flow, added interface conformance assertions for sergice and hardened validator dependency handling so alternate implemntnations cant trigger panics or rely on stronger preconditions.

* Applying liskov substitution. Default clock implementation, guard repo fetches against nil markets and invalid ids, normalize search and overview helpers to handle nil slices.

* Removed panic on missing seeds, getSeeds now supplies and defaults, PositionCalculator toleratoes nil dependencies by falling back to default probability / payout models, preventing panics from zero value calculators.

* Added LSP safeguards, service entrypoints against missing dependencies, ComputeUserFinancials and ComputeSystemMetrics, service construction seeds default strategies via ensureStrategyDefaults, interface assertions, Gorm repo methods now fail fast with descriptive error.

---------

Co-authored-by: Patrick W. Delaney <patrick@Mac.hsd1.tx.comcast.net>

pwdel committed:
Added MarketReadRepository, MarketWriteRepository, MarketPositionRepository, MarketBetRepository and update resolution policy and helpers. (#613)

Co-authored-by: Patrick W. Delaney <patrick@Mac.hsd1.tx.comcast.net>

pwdel committed:
DIP (#614)

* Updating.

* Updating.

* Updating.

---------

Co-authored-by: Patrick W. Delaney <patrick@Mac.hsd1.tx.comcast.net>

pwdel commented:
This was completed in a different PR https://github.com/openpredictionmarkets/socialpredict/pull/657

pwdel commented:
https://github.com/openpredictionmarkets/socialpredict/pull/657
